crates/bevy_pbr/src/render/mesh_functions.wgsl

```diff
@@ -0,0 +1,36 @@
+#define_import_path bevy_pbr::mesh_functions
+
+fn mesh_position_local_to_world(model: mat4x4<f32>, vertex_position: vec4<f32>) -> vec4<f32> {
```

Comment:
I'm pretty sure I've endorsed this idea in the past, but having thought about it a bit more, I think I'd like to push back on this general pattern a bit. Especially for the functions that just alias multiplication. Is introducing special-case methods for those operations "worth it"?

On the one hand, they give descriptive (and consistent) names to the operations (mesh_position_local_to_world, mesh_position_world_to_clip), etc.

On the other hand, teaching people to use this:
`mesh_position_local_to_world(model, position)`
over this:
`model * position`

Feels wrong to me. It is abstracting out a fundamental concept (vec/matrix multiplication) in favor of a custom case-specific concept. Worse, that case-specific concept / naming doesn't hold if you pass a non-model matrix in (or non mesh position). And on top of that, it requires you to pull in an import that you might not otherwise need.

I think we should only provide helper functions that actually reduce boilerplate and/or fully abstract out concepts:
* Functions that abstract out multiple fundamental operations, such as `mesh_position_local_to_clip` replacing a "world position" operation followed by a "world to clip" operation.
* Functions that fully abstract out bindings. On that topic, I think the UX of  `mesh_position_local_to_clip` would benefit from abstracting out the model matrix and taking a single vec as input (like Unity does in their shaders). That makes it dead-simple to call and better justifies the abstraction cost. I think I see why we aren't doing that here (avoiding double-calculating mesh skinning?). But that seems like a technical hurdle worth solving (maybe not in this pr, but at some point).

Reply:
Note a complete response, just a quick one. The value I see for the local to world case is that it creates consistency and mental clarity that you’re doing it in the right way according to the engine regardless of whether your calculating the world or clip space versions. You could argue that someone should know that projection * inverse view * model * local position = clip position, and I would agree. But when doing a depth prepass (which I am expecting we will at some point) and using the equal depth comparison function in the main pass, float precision of the calculations will matter and I wouldn’t expect someone to know that. And then if you don’t have a function for local to world but do for local to clip, then there’s some inconsistency in the story. There’s nothing stopping people just using the matrices in the bindings, but if code examples use these functions then people will be encouraged to use them. When I move the mesh uniform into an instance buffer for performance reasons, you anyway won’t be able to just refer to mesh.model and will need to reconstruct the model matrix.

Maybe we can use some ‘global variables’ to cache calculation of matrices somehow so that you only pass in the vector to be transformed…? I don’t know if wgsl would allow that.

Reply:
If we actually abstracted out the mesh matrix (enabling things like abstracting out the instance buffer access in the future) id more convinced. I still think as it stands the "consistency" of `mesh_position_local_to_world` doesn't justify its complexity cost. As written, it doesn't actually help us with abstracting out things like "mesh matrix sources". And it makes it too easy to use the function in the "wrong" way (by passing in the wrong matrix).

Reply:
Just verified that this is possible:

```
var<private> MESH_MODEL: mat4x4<f32>;

fn mesh_position_local_to_world(vertex_position: vec4<f32>) -> vec4<f32> {
    return MESH_MODEL * vertex_position;
}

fn update_mesh_model(vertex: Vertex) {
    #ifdef SKINNED
    skin_model(vertex.joint_indices, vertex.joint_weights);
    #else
    MESH_MODEL = mesh.model;
    #endif
}
```

It would require calling `update_mesh_model(vertex)` at the start of the vertex shader. But it would also provide the benefit of fully abstracting out skinning for most use cases.

Another downside is that this requires doing the `#import bevy_pbr::mesh_functions` call _after_ defining Vertex.

Reply:
We could also do lazy calculations with a second `var<private> MESH_MODEL_CALCULATED: bool;`
But I'm hesitant to add runtime calculations for what should really be static.

Reply:
We could also use the godot approach: treat user-defined vertex + fragment functions as "internal":

```rust
[[stage(vertex)]]
fn internal_vertex(vertex: Vertex) -> VertexOutput {
  /* do setup work here, such as setting MESH_MODEL */
  return vertex(vertex);
}

// user defines this, which will implicitly get injected in to a template containing `internal_vertex` / whatever else is needed
fn vertex(vertex: Vertex) -> VertexOutput {
  // user is free to assume MESH_MODEL is already valid and
  // call things like mesh_position_local_to_world(vertex.position)
}
```

Definitely more abstraction (and more "fixed"). Just throwing out options to get the juices flowing.
Also, in the interest of moving forward on "callable pbr", im happy to table this, as I'm encouraging some serious scope creep here, and your changes are non-breaking / optional.

But I still feel relatively strongly about the core principles outlined above in the bullet points of this message: https://github.com/bevyengine/bevy/pull/4901#discussion_r891774566

Reply:
It maybe doesn’t need to be global. But I don’t know if you like that. I was going to try a ptr with function scope.